File: src/grids/arakawa.jl
"""
Arakawa staggering accessor runtime.

Arakawa staggering is a *transform* over an underlying base grid. Given a base
grid (Cartesian, lat-lon, …) and a stagger label (A/B/C/D/E), the arakawa
runtime provides on-demand accessors for cell-center, u-face, v-face, and
corner locations per the staggering convention.

Per GRIDS_API.md §0 correction: the `.esm` lowering is a small declarative
config (family + base-grid ref + stagger + dimensions + extents). Geometry
is derived from that config by the accessors here, not serialized as a blob.

Staggering conventions (2-D, horizontal):

- A: h, u, v colocated at cell centers.
- B: h at cell centers; u, v colocated at corners.
- C: h at cell centers; u at u-faces (east-west); v at v-faces (north-south).
- D: h at cell centers; u at v-faces; v at u-faces (swapped from C).
- E: rotated B-grid; topologically equivalent to B with a 45° rotation flag
  in the lowered config.
"""

@enum ArakawaStagger::UInt8 ArakawaA ArakawaB ArakawaC ArakawaD ArakawaE

# Reuse VarLocation from staggering.jl (loaded before this file via
# EarthSciDiscretizations.jl include order).

# Variable-to-location table per stagger. Returns (h_loc, u_loc, v_loc).
function arakawa_variable_locations(s::ArakawaStagger)
    return s === ArakawaA ? (CellCenter, CellCenter, CellCenter) :
        s === ArakawaB ? (CellCenter, Corner, Corner) :
        s === ArakawaC ? (CellCenter, UEdge, VEdge) :
        s === ArakawaD ? (CellCenter, VEdge, UEdge) :
        (CellCenter, Corner, Corner)  # ArakawaE: like B, rotated 45°
end

# Location shape in (ni, nj) on a base grid of (nx, ny) interior cells.
function arakawa_location_shape(loc::VarLocation, nx::Int, ny::Int)
    return loc === CellCenter ? (nx, ny) :
        loc === UEdge ? (nx + 1, ny) :
        loc === VEdge ? (nx, ny + 1) :
        (nx + 1, ny + 1)   # Corner
end

# ---------------------------------------------------------------------------
# Base grid abstraction
#
# Phase 1 (cartesian) and Phase 2/3 (lat-lon) grid types are not yet on main.
# The arakawa runtime is parameterised over an `ArakawaBaseGrid`; a minimal
# `CartesianBase` is provided here so the accessors are exercisable today.
# Future Phase 1/2/3 grid structs can subtype `ArakawaBaseGrid` and supply
# the same four primitives (cell-center, x-edge, y-edge, corner coordinates
# plus cell counts) to plug in without touching this file.
# ---------------------------------------------------------------------------

abstract type ArakawaBaseGrid end

struct CartesianBase{T} <: ArakawaBaseGrid
    xlo::T
    xhi::T
    ylo::T
    yhi::T
    nx::Int
    ny::Int
end

function CartesianBase(; xlo::Real, xhi::Real, ylo::Real, yhi::Real, nx::Int, ny::Int)
    nx > 0 || throw(DomainError(nx, "nx must be positive"))
    ny > 0 || throw(DomainError(ny, "ny must be positive"))
    xhi > xlo || throw(DomainError(xhi, "xhi must be greater than xlo"))
    yhi > ylo || throw(DomainError(yhi, "yhi must be greater than ylo"))
    T = promote_type(typeof(xlo), typeof(xhi), typeof(ylo), typeof(yhi))
    return CartesianBase{T}(T(xlo), T(xhi), T(ylo), T(yhi), nx, ny)
end

arakawa_nx(b::CartesianBase) = b.nx
arakawa_ny(b::CartesianBase) = b.ny
arakawa_dx(b::CartesianBase{T}) where {T} = (b.xhi - b.xlo) / T(b.nx)
arakawa_dy(b::CartesianBase{T}) where {T} = (b.yhi - b.ylo) / T(b.ny)

function arakawa_cell_center(b::CartesianBase{T}, i::Int, j::Int) where {T}
    return (
        b.xlo + (T(i) - T(0.5)) * arakawa_dx(b),
        b.ylo + (T(j) - T(0.5)) * arakawa_dy(b),
    )
end

function arakawa_x_edge(b::CartesianBase{T}, i::Int, j::Int) where {T}
    return (
        b.xlo + T(i - 1) * arakawa_dx(b),
        b.ylo + (T(j) - T(0.5)) * arakawa_dy(b),
    )
end

function arakawa_y_edge(b::CartesianBase{T}, i::Int, j::Int) where {T}
    return (
        b.xlo + (T(i) - T(0.5)) * arakawa_dx(b),
        b.ylo + T(j - 1) * arakawa_dy(b),
    )
end

function arakawa_corner(b::CartesianBase{T}, i::Int, j::Int) where {T}
    return (
        b.xlo + T(i - 1) * arakawa_dx(b),
        b.ylo + T(j - 1) * arakawa_dy(b),
    )
end

_arakawa_base_esm(b::CartesianBase) = Dict{String, Any}(
    "family" => "cartesian",
    "nx" => b.nx,
    "ny" => b.ny,
    "extent" => [[b.xlo, b.ylo], [b.xhi, b.yhi]],
)

# ---------------------------------------------------------------------------
# ArakawaGrid: a base grid + stagger label
# ---------------------------------------------------------------------------

struct ArakawaGrid{T, B <: ArakawaBaseGrid} <: AbstractStaggeredGrid
    base::B
    stagger::ArakawaStagger
    ghosts::Int
end

function ArakawaGrid(
        base::B, stagger::ArakawaStagger;
        ghosts::Int = 0, dtype::Type{T} = Float64
    ) where {T, B <: ArakawaBaseGrid}
    ghosts >= 0 || throw(DomainError(ghosts, "ghosts must be non-negative"))
    return ArakawaGrid{T, B}(base, stagger, ghosts)
end

Base.eltype(::Type{ArakawaGrid{T, B}}) where {T, B} = T
Base.eltype(g::ArakawaGrid) = eltype(typeof(g))

arakawa_nx(g::ArakawaGrid) = arakawa_nx(g.base)
arakawa_ny(g::ArakawaGrid) = arakawa_ny(g.base)

# Shape of a specific location on this grid (interior cells only; ghosts
# are reported separately via `ghosts` and are not expanded here).
arakawa_shape(g::ArakawaGrid, loc::VarLocation) =
    arakawa_location_shape(loc, arakawa_nx(g), arakawa_ny(g))

# Shape of the location where each named variable lives (h, u, v).
function variable_shape(g::ArakawaGrid, var::Symbol)
    h_loc, u_loc, v_loc = arakawa_variable_locations(g.stagger)
    loc = var === :h ? h_loc :
        var === :u ? u_loc :
        var === :v ? v_loc :
        throw(ArgumentError("Unknown arakawa variable $var; expected :h, :u, or :v"))
    return arakawa_shape(g, loc)
end

# ---------------------------------------------------------------------------
# Accessors — DERIVE geometry from the declaration (mayor's correction).
# ---------------------------------------------------------------------------

"""
    cell_centers(g::ArakawaGrid, i::Int, j::Int) -> (x, y)

Coordinate of cell-centre `(i, j)` (`1`-based). `(i, j)` must satisfy
`1 ≤ i ≤ nx` and `1 ≤ j ≤ ny`.
"""
function cell_centers(g::ArakawaGrid, i::Int, j::Int)
    _check_bounds(g, CellCenter, i, j)
    return arakawa_cell_center(g.base, i, j)
end

"""
    u_face(g::ArakawaGrid, i::Int, j::Int) -> (x, y)

Coordinate of the location where `u` lives at index `(i, j)` for this
stagger. Valid index range depends on the stagger (see `variable_shape`).
"""
function u_face(g::ArakawaGrid, i::Int, j::Int)
    _, u_loc, _ = arakawa_variable_locations(g.stagger)
    _check_bounds(g, u_loc, i, j)
    return _coord_at(g, u_loc, i, j)
end

"""
    v_face(g::ArakawaGrid, i::Int, j::Int) -> (x, y)

Coordinate of the location where `v` lives at index `(i, j)` for this
stagger.
"""
function v_face(g::ArakawaGrid, i::Int, j::Int)
    _, _, v_loc = arakawa_variable_locations(g.stagger)
    _check_bounds(g, v_loc, i, j)
    return _coord_at(g, v_loc, i, j)
end

"""
    corners(g::ArakawaGrid, i::Int, j::Int) -> (x, y)

Coordinate of cell-corner `(i, j)` (`1 ≤ i ≤ nx+1`, `1 ≤ j ≤ ny+1`).
"""
function corners(g::ArakawaGrid, i::Int, j::Int)
    _check_bounds(g, Corner, i, j)
    return arakawa_corner(g.base, i, j)
end

function _coord_at(g::ArakawaGrid, loc::VarLocation, i::Int, j::Int)
    return loc === CellCenter ? arakawa_cell_center(g.base, i, j) :
        loc === UEdge ? arakawa_x_edge(g.base, i, j) :
        loc === VEdge ? arakawa_y_edge(g.base, i, j) :
        arakawa_corner(g.base, i, j)
end

function _check_bounds(g::ArakawaGrid, loc::VarLocation, i::Int, j::Int)
    ni, nj = arakawa_shape(g, loc)
    return (1 <= i <= ni && 1 <= j <= nj) ||
        throw(DomainError((i, j), "index out of bounds for location $loc with shape ($ni, $nj)"))
end

"""
    neighbors(g::ArakawaGrid, loc::VarLocation, i::Int, j::Int)
        -> (west, east, south, north)

Return the four axial neighbours of `(loc, i, j)` as `(loc, i, j)` tuples, or
`nothing` at domain boundaries. Does not cross ghost cells (ghost halos are
reported via `g.ghosts` but not expanded into the index range here).
"""
function neighbors(g::ArakawaGrid, loc::VarLocation, i::Int, j::Int)
    _check_bounds(g, loc, i, j)
    ni, nj = arakawa_shape(g, loc)
    west = i > 1 ? (loc, i - 1, j) : nothing
    east = i < ni ? (loc, i + 1, j) : nothing
    south = j > 1 ? (loc, i, j - 1) : nothing
    north = j < nj ? (loc, i, j + 1) : nothing
    return (west, east, south, north)
end

neighbors(g::ArakawaGrid, i::Int, j::Int) = neighbors(g, CellCenter, i, j)

"""
    metric_eval(g::ArakawaGrid, name::Symbol, i::Int, j::Int) -> scalar

Return the named metric at cell `(i, j)`. Supported names:

- `:dx`   — zonal spacing at `(i, j)`
- `:dy`   — meridional spacing at `(i, j)`
- `:area` — cell-centre area / volume

For `CartesianBase`, metrics are uniform and the `(i, j)` arguments are
ignored (bounds are still checked).
"""
function metric_eval(g::ArakawaGrid{T}, name::Symbol, i::Int, j::Int) where {T}
    _check_bounds(g, CellCenter, i, j)
    dx = T(arakawa_dx(g.base))
    dy = T(arakawa_dy(g.base))
    return if name === :dx
        dx
    elseif name === :dy
        dy
    elseif name === :area
        dx * dy
    else
        throw(ArgumentError("Unknown metric $name; expected :dx, :dy, or :area"))
    end
end

# ---------------------------------------------------------------------------
# .esm lowering: declarative config per §6 schema + the §0 correction.
# ---------------------------------------------------------------------------

"""
    to_esm(g::ArakawaGrid) -> Dict{String,Any}

Lower the grid to a §6-schema-valid `.esm` declaration. The result is a
small config — `family`, `stagger`, `base`, `dtype`, `ghosts`, `n_cells`,
`topology` — without any inline geometry arrays.
"""
function to_esm(g::ArakawaGrid{T}) where {T}
    return Dict{String, Any}(
        "family" => "arakawa",
        "dtype" => _dtype_string(T),
        "topology" => "block_structured",
        "ghosts" => g.ghosts,
        "n_cells" => arakawa_nx(g) * arakawa_ny(g),
        "stagger" => _stagger_string(g.stagger),
        "rotated" => g.stagger === ArakawaE,
        "base" => _arakawa_base_esm(g.base),
    )
end

function _stagger_string(s::ArakawaStagger)
    return s === ArakawaA ? "A" :
        s === ArakawaB ? "B" :
        s === ArakawaC ? "C" :
        s === ArakawaD ? "D" : "E"
end

_dtype_string(::Type{Float64}) = "float64"
_dtype_string(::Type{Float32}) = "float32"
_dtype_string(::Type{T}) where {T} = string(T)

# ---------------------------------------------------------------------------
# ESS Grid trait — Tier C + Tier S (staggered horizontal mesh)
#
# Cell-centered Tier-C arrays are flat row-major over the (nx, ny) interior
# (slow axis = j). Face / corner accessors stay scalar — they are
# stagger-specific and live above as `u_face`, `v_face`, `corners`.
# ---------------------------------------------------------------------------

n_dims(::ArakawaGrid) = 2
axis_names(::ArakawaGrid) = (:x, :y)
n_cells(g::ArakawaGrid) = arakawa_nx(g) * arakawa_ny(g)

function _arakawa_axis_idx(::ArakawaGrid, axis::Symbol)
    axis === :x && return 1
    axis === :y && return 2
    throw(ArgumentError("arakawa: unknown axis :$axis (expected :x or :y)"))
end

function cell_centers(g::ArakawaGrid{T}, axis::Symbol) where {T}
    d = _arakawa_axis_idx(g, axis)
    return _grid_memo!(g, (:cell_centers, axis)) do
        nx = arakawa_nx(g); ny = arakawa_ny(g)
        out = Vector{T}(undef, nx * ny)
        @inbounds for j in 1:ny, i in 1:nx
            xy = arakawa_cell_center(g.base, i, j)
            out[(j - 1) * nx + i] = T(xy[d])
        end
        return out
    end
end

function cell_widths(g::ArakawaGrid{T}, axis::Symbol) where {T}
    d = _arakawa_axis_idx(g, axis)
    return _grid_memo!(g, (:cell_widths, axis)) do
        nx = arakawa_nx(g); ny = arakawa_ny(g)
        w = d == 1 ? T(arakawa_dx(g.base)) : T(arakawa_dy(g.base))
        return fill(w, nx * ny)
    end
end

function cell_volume(g::ArakawaGrid{T}) where {T}
    return _grid_memo!(g, :cell_volume) do
        nx = arakawa_nx(g); ny = arakawa_ny(g)
        a = T(arakawa_dx(g.base)) * T(arakawa_dy(g.base))
        return fill(a, nx * ny)
    end
end

function neighbor_indices(g::ArakawaGrid, axis::Symbol, offset::Int)
    d = _arakawa_axis_idx(g, axis)
    return _grid_memo!(g, (:neighbor_indices, axis, offset)) do
        nx = arakawa_nx(g); ny = arakawa_ny(g)
        out = Vector{Int}(undef, nx * ny)
        @inbounds for j in 1:ny, i in 1:nx
            ii = d == 1 ? i + offset : i
            jj = d == 2 ? j + offset : j
            k = (j - 1) * nx + i
            out[k] = (1 <= ii <= nx && 1 <= jj <= ny) ?
                ((jj - 1) * nx + ii) : 0
        end
        return out
    end
end

function boundary_mask(g::ArakawaGrid, axis::Symbol, side::Symbol)
    d = _arakawa_axis_idx(g, axis)
    side in (:lower, :upper) ||
        throw(ArgumentError("arakawa: side must be :lower or :upper; got :$side"))
    return _grid_memo!(g, (:boundary_mask, axis, side)) do
        nx = arakawa_nx(g); ny = arakawa_ny(g)
        out = falses(nx * ny)
        target = side === :lower ? 1 : (d == 1 ? nx : ny)
        @inbounds for j in 1:ny, i in 1:nx
            v = d == 1 ? i : j
            out[(j - 1) * nx + i] = v == target
        end
        return out
    end
end

File: typescript/src/grids/arakawa.ts
/**
 * Arakawa staggering accessor runtime.
 *
 * Conforms to the cross-binding contract in `docs/GRIDS_API.md` §2.6, §3.4, §7.
 *
 * Arakawa staggering is a *transform* over an underlying base grid. Given a
 * base grid (Cartesian or lat-lon) plus a stagger label (A/B/C/D/E), the
 * runtime provides on-demand accessors for cell centers, u-face, v-face, and
 * corner locations per the staggering convention.
 *
 * Per the 2026-04-20 scope correction, `toESM()` emits a small declarative
 * config (family + base-grid ref + stagger + dimensions + extents), NOT a
 * serialized geometry blob. Geometry is derived from that config by the
 * accessors on demand.
 *
 * Stagger conventions (2-D, horizontal):
 *
 *   - A: h, u, v colocated at cell centers.
 *   - B: h at cell centers; u, v colocated at corners.
 *   - C: h at cell centers; u at u-faces (east-west); v at v-faces (north-south).
 *   - D: h at cell centers; u at v-faces; v at u-faces (swapped from C).
 *   - E: rotated B-grid; topologically equivalent to B with a 45° rotation flag
 *     in the lowered config.
 *
 * Mirrors the Julia reference binding (`src/grids/arakawa.jl`) and the Rust
 * binding (`rust/src/grids/arakawa.rs`). Indices are 0-based in TS, matching
 * `cartesian.ts` and `cubed_sphere.ts`.
 */

import type { Dtype, Grid } from "./types.js";

const ARAKAWA_API_VERSION = "1.0.0";
const ARAKAWA_BINDING_VERSION = "1.0.0";

// ---------------------------------------------------------------------------
// Stagger + Location + Variable vocabularies (snake_case on the wire).
// ---------------------------------------------------------------------------

export type ArakawaStagger = "A" | "B" | "C" | "D" | "E";

export type ArakawaLocation = "cell_center" | "u_edge" | "v_edge" | "corner";

export type ArakawaVariable = "h" | "u" | "v";

export type ArakawaMetricName = "dx" | "dy" | "area";

const STAGGERS: readonly ArakawaStagger[] = ["A", "B", "C", "D", "E"];
const LOCATIONS: readonly ArakawaLocation[] = [
  "cell_center",
  "u_edge",
  "v_edge",
  "corner",
];
const VARIABLES: readonly ArakawaVariable[] = ["h", "u", "v"];
const METRICS: readonly ArakawaMetricName[] = ["dx", "dy", "area"];

/** Per-stagger `(h_loc, u_loc, v_loc)` variable-location triple. */
export function variableLocations(
  s: ArakawaStagger,
): [ArakawaLocation, ArakawaLocation, ArakawaLocation] {
  switch (s) {
    case "A":
      return ["cell_center", "cell_center", "cell_center"];
    case "B":
      return ["cell_center", "corner", "corner"];
    case "C":
      return ["cell_center", "u_edge", "v_edge"];
    case "D":
      return ["cell_center", "v_edge", "u_edge"];
    case "E":
      // E: topologically like B; 45° rotation carried in `rotated` flag.
      return ["cell_center", "corner", "corner"];
  }
}

/** Shape `(ni, nj)` of `loc` on a base grid of `(nx, ny)` interior cells. */
export function locationShape(
  loc: ArakawaLocation,
  nx: number,
  ny: number,
): [number, number] {
  switch (loc) {
    case "cell_center":
      return [nx, ny];
    case "u_edge":
      return [nx + 1, ny];
    case "v_edge":
      return [nx, ny + 1];
    case "corner":
      return [nx + 1, ny + 1];
  }
}

// ---------------------------------------------------------------------------
// Base grid abstraction.
//
// Phase 1 (cartesian) and Phase 2/3 (lat-lon) grid types exist but have their
// own options shapes. Arakawa is parameterised over an `ArakawaBaseGrid`; a
// minimal `CartesianBase` is provided here so the accessors are exercisable
// today. Future base-grid families implement the same interface and plug in
// without touching this file.
// ---------------------------------------------------------------------------

/** Primitives every arakawa base grid must expose. */
export interface ArakawaBaseGrid {
  /** Interior cells along x. */
  readonly nx: number;
  /** Interior cells along y. */
  readonly ny: number;
  /** Uniform cell spacing in x. */
  dx(): number;
  /** Uniform cell spacing in y. */
  dy(): number;
  /** `(x, y)` of cell center `(i, j)` (0-based). */
  cellCenter(i: number, j: number): [number, number];
  /** `(x, y)` of the u-face location at `(i, j)` (0-based, `0..nx × 0..ny-1`). */
  xEdge(i: number, j: number): [number, number];
  /** `(x, y)` of the v-face location at `(i, j)` (0-based, `0..nx-1 × 0..ny`). */
  yEdge(i: number, j: number): [number, number];
  /** `(x, y)` of the cell corner at `(i, j)` (0-based, `0..nx × 0..ny`). */
  corner(i: number, j: number): [number, number];
  /** Declarative `.esm`-shaped summary of the base grid. */
  toESM(): object;
}

export interface CartesianBaseOpts {
  xlo: number;
  xhi: number;
  ylo: number;
  yhi: number;
  nx: number;
  ny: number;
}

export interface CartesianBase extends ArakawaBaseGrid {
  readonly family: "cartesian";
  readonly xlo: number;
  readonly xhi: number;
  readonly ylo: number;
  readonly yhi: number;
}

function isPositiveInteger(v: unknown): v is number {
  return typeof v === "number" && Number.isInteger(v) && v >= 1;
}

function isFiniteNumber(v: unknown): v is number {
  return typeof v === "number" && Number.isFinite(v);
}

/** Build a validated uniform Cartesian base grid. */
export function cartesianBase(opts: CartesianBaseOpts): CartesianBase {
  if (opts === undefined || opts === null) {
    throw new TypeError("cartesianBase: options object is required");
  }
  const { xlo, xhi, ylo, yhi, nx, ny } = opts;
  if (!isPositiveInteger(nx)) {
    throw new RangeError(
      `cartesianBase: 'nx' must be a positive integer (got ${String(nx)})`,
    );
  }
  if (!isPositiveInteger(ny)) {
    throw new RangeError(
      `cartesianBase: 'ny' must be a positive integer (got ${String(ny)})`,
    );
  }
  if (!isFiniteNumber(xlo) || !isFiniteNumber(xhi)) {
    throw new RangeError(
      `cartesianBase: 'xlo' and 'xhi' must be finite numbers (got xlo=${String(xlo)}, xhi=${String(xhi)})`,
    );
  }
  if (!isFiniteNumber(ylo) || !isFiniteNumber(yhi)) {
    throw new RangeError(
      `cartesianBase: 'ylo' and 'yhi' must be finite numbers (got ylo=${String(ylo)}, yhi=${String(yhi)})`,
    );
  }
  if (!(xhi > xlo)) {
    throw new RangeError(
      `cartesianBase: 'xhi' must be > 'xlo' (got xlo=${xlo}, xhi=${xhi})`,
    );
  }
  if (!(yhi > ylo)) {
    throw new RangeError(
      `cartesianBase: 'yhi' must be > 'ylo' (got ylo=${ylo}, yhi=${yhi})`,
    );
  }

  const dxVal = (xhi - xlo) / nx;
  const dyVal = (yhi - ylo) / ny;

  const base: CartesianBase = {
    family: "cartesian",
    xlo,
    xhi,
    ylo,
    yhi,
    nx,
    ny,
    dx(): number {
      return dxVal;
    },
    dy(): number {
      return dyVal;
    },
    cellCenter(i: number, j: number): [number, number] {
      return [xlo + (i + 0.5) * dxVal, ylo + (j + 0.5) * dyVal];
    },
    xEdge(i: number, j: number): [number, number] {
      return [xlo + i * dxVal, ylo + (j + 0.5) * dyVal];
    },
    yEdge(i: number, j: number): [number, number] {
      return [xlo + (i + 0.5) * dxVal, ylo + j * dyVal];
    },
    corner(i: number, j: number): [number, number] {
      return [xlo + i * dxVal, ylo + j * dyVal];
    },
    toESM(): object {
      return {
        family: "cartesian",
        nx,
        ny,
        extent: [
          [xlo, ylo],
          [xhi, yhi],
        ],
      };
    },
  };
  return base;
}

// ---------------------------------------------------------------------------
// ArakawaGrid interface + options.
// ---------------------------------------------------------------------------

export interface ArakawaOpts {
  base: ArakawaBaseGrid;
  stagger: ArakawaStagger;
  dtype?: Dtype;
  ghosts?: number;
}

export interface ArakawaNeighbors {
  readonly west: [number, number] | null;
  readonly east: [number, number] | null;
  readonly south: [number, number] | null;
  readonly north: [number, number] | null;
}

export interface ArakawaGrid extends Grid {
  readonly family: "arakawa";
  readonly topology: "block_structured";
  readonly stagger: ArakawaStagger;
  readonly rotated: boolean;
  readonly ghosts: number;
  readonly base: ArakawaBaseGrid;
  readonly nx: number;
  readonly ny: number;
  readonly n_cells: number;

  /** Shape `(ni, nj)` of the named location on this grid. */
  location_shape(loc: ArakawaLocation): [number, number];
  /** Location where `var` lives under this grid's stagger. */
  variable_location(variable: ArakawaVariable): ArakawaLocation;
  /** Shape of the location where `var` lives under this stagger. */
  variable_shape(variable: ArakawaVariable): [number, number];

  /** `(x, y)` of cell center `(i, j)` (0-based). */
  cell_centers(i: number, j: number): [number, number];
  /** `(x, y)` of the u-variable location at `(i, j)` under this stagger. */
  u_face(i: number, j: number): [number, number];
  /** `(x, y)` of the v-variable location at `(i, j)` under this stagger. */
  v_face(i: number, j: number): [number, number];
  /** `(x, y)` of cell corner `(i, j)` (0-based, `0..nx × 0..ny`). */
  corners(i: number, j: number): [number, number];
  /** Generic `(x, y)` of `(loc, i, j)`. */
  coord(loc: ArakawaLocation, i: number, j: number): [number, number];

  /**
   * Four axial neighbours of `(loc, i, j)` as `{west, east, south, north}`,
   * each `[i', j']` or `null` at a domain boundary. Stays on the same
   * location (no cross-location stepping — that's the caller's job per
   * stagger semantics).
   */
  neighbors(loc: ArakawaLocation, i: number, j: number): ArakawaNeighbors;

  /** Evaluate a named metric at cell `(i, j)`. */
  metric_eval(name: ArakawaMetricName, i: number, j: number): number;
}

// ---------------------------------------------------------------------------
// Generator.
// ---------------------------------------------------------------------------

function checkBounds(
  loc: ArakawaLocation,
  i: number,
  j: number,
  nx: number,
  ny: number,
): void {
  const [ni, nj] = locationShape(loc, nx, ny);
  if (!Number.isInteger(i) || i < 0 || i >= ni) {
    throw new RangeError(
      `arakawa: i=${String(i)} out of range [0, ${ni}) for location '${loc}'`,
    );
  }
  if (!Number.isInteger(j) || j < 0 || j >= nj) {
    throw new RangeError(
      `arakawa: j=${String(j)} out of range [0, ${nj}) for location '${loc}'`,
    );
  }
}

function coordAt(
  base: ArakawaBaseGrid,
  loc: ArakawaLocation,
  i: number,
  j: number,
): [number, number] {
  switch (loc) {
    case "cell_center":
      return base.cellCenter(i, j);
    case "u_edge":
      return base.xEdge(i, j);
    case "v_edge":
      return base.yEdge(i, j);
    case "corner":
      return base.corner(i, j);
  }
}

function isBaseGridLike(v: unknown): v is ArakawaBaseGrid {
  if (v === null || typeof v !== "object") return false;
  const b = v as Record<string, unknown>;
  return (
    typeof b.nx === "number" &&
    typeof b.ny === "number" &&
    typeof b.dx === "function" &&
    typeof b.dy === "function" &&
    typeof b.cellCenter === "function" &&
    typeof b.xEdge === "function" &&
    typeof b.yEdge === "function" &&
    typeof b.corner === "function" &&
    typeof b.toESM === "function"
  );
}

export function arakawa(opts: ArakawaOpts): ArakawaGrid {
  if (opts === undefined || opts === null) {
    throw new TypeError("arakawa: options object is required");
  }

  if (!isBaseGridLike(opts.base)) {
    throw new TypeError(
      "arakawa: missing or invalid required option 'base' (expected an ArakawaBaseGrid)",
    );
  }
  if (opts.stagger === undefined) {
    throw new TypeError("arakawa: missing required option 'stagger'");
  }
  if (!STAGGERS.includes(opts.stagger)) {
    throw new RangeError(
      `arakawa: 'stagger' must be one of ${STAGGERS.join("/")} (got ${String(opts.stagger)})`,
    );
  }

  const dtype: Dtype = opts.dtype ?? "float64";
  if (dtype !== "float64" && dtype !== "float32") {
    throw new RangeError(
      `arakawa: 'dtype' must be 'float64' or 'float32' (got ${String(dtype)})`,
    );
  }

  const ghosts = opts.ghosts ?? 0;
  if (!Number.isInteger(ghosts) || ghosts < 0) {
    throw new RangeError(
      `arakawa: 'ghosts' must be a non-negative integer (got ${String(ghosts)})`,
    );
  }

  const base = opts.base;
  if (!isPositiveInteger(base.nx)) {
    throw new RangeError(
      `arakawa: base.nx must be a positive integer (got ${String(base.nx)})`,
    );
  }
  if (!isPositiveInteger(base.ny)) {
    throw new RangeError(
      `arakawa: base.ny must be a positive integer (got ${String(base.ny)})`,
    );
  }

  const stagger: ArakawaStagger = opts.stagger;
  const rotated = stagger === "E";
  const [hLoc, uLoc, vLoc] = variableLocations(stagger);

  const grid: ArakawaGrid = {
    family: "arakawa",
    topology: "block_structured",
    dtype,
    stagger,
    rotated,
    ghosts,
    base,
    get nx(): number {
      return base.nx;
    },
    get ny(): number {
      return base.ny;
    },
    get n_cells(): number {
      return base.nx * base.ny;
    },

    location_shape(loc: ArakawaLocation): [number, number] {
      if (!LOCATIONS.includes(loc)) {
        throw new RangeError(
          `arakawa: unknown location '${String(loc)}'; expected one of ${LOCATIONS.join("/")}`,
        );
      }
      return locationShape(loc, base.nx, base.ny);
    },

    variable_location(variable: ArakawaVariable): ArakawaLocation {
      switch (variable) {
        case "h":
          return hLoc;
        case "u":
          return uLoc;
        case "v":
          return vLoc;
        default:
          throw new RangeError(
            `arakawa: unknown variable '${String(variable)}'; expected one of ${VARIABLES.join("/")}`,
          );
      }
    },

    variable_shape(variable: ArakawaVariable): [number, number] {
      return locationShape(
        this.variable_location(variable),
        base.nx,
        base.ny,
      );
    },

    cell_centers(i: number, j: number): [number, number] {
      checkBounds("cell_center", i, j, base.nx, base.ny);
      return base.cellCenter(i, j);
    },

    u_face(i: number, j: number): [number, number] {
      checkBounds(uLoc, i, j, base.nx, base.ny);
      return coordAt(base, uLoc, i, j);
    },

    v_face(i: number, j: number): [number, number] {
      checkBounds(vLoc, i, j, base.nx, base.ny);
      return coordAt(base, vLoc, i, j);
    },

    corners(i: number, j: number): [number, number] {
      checkBounds("corner", i, j, base.nx, base.ny);
      return base.corner(i, j);
    },

    coord(loc: ArakawaLocation, i: number, j: number): [number, number] {
      if (!LOCATIONS.includes(loc)) {
        throw new RangeError(
          `arakawa: unknown location '${String(loc)}'; expected one of ${LOCATIONS.join("/")}`,
        );
      }
      checkBounds(loc, i, j, base.nx, base.ny);
      return coordAt(base, loc, i, j);
    },

    neighbors(
      loc: ArakawaLocation,
      i: number,
      j: number,
    ): ArakawaNeighbors {
      if (!LOCATIONS.includes(loc)) {
        throw new RangeError(
          `arakawa: unknown location '${String(loc)}'; expected one of ${LOCATIONS.join("/")}`,
        );
      }
      checkBounds(loc, i, j, base.nx, base.ny);
      const [ni, nj] = locationShape(loc, base.nx, base.ny);
      return {
        west: i > 0 ? [i - 1, j] : null,
        east: i + 1 < ni ? [i + 1, j] : null,
        south: j > 0 ? [i, j - 1] : null,
        north: j + 1 < nj ? [i, j + 1] : null,
      };
    },

    metric_eval(name: ArakawaMetricName, i: number, j: number): number {
      if (!METRICS.includes(name)) {
        throw new RangeError(
          `arakawa: unknown metric '${String(name)}'; expected one of ${METRICS.join("/")}`,
        );
      }
      checkBounds("cell_center", i, j, base.nx, base.ny);
      const dxV = base.dx();
      const dyV = base.dy();
      switch (name) {
        case "dx":
          return dxV;
        case "dy":
          return dyV;
        case "area":
          return dxV * dyV;
      }
    },

    toESM(): object {
      return {
        family: "arakawa",
        version: ARAKAWA_API_VERSION,
        dtype,
        topology: "block_structured",
        ghosts,
        n_cells: base.nx * base.ny,
        stagger,
        rotated,
        base: base.toESM(),
        provenance: {
          binding: "typescript",
          binding_version: ARAKAWA_BINDING_VERSION,
          api_version: ARAKAWA_API_VERSION,
          platform: "js",
          runtime: "node",
          math_lib: "libm-js",
          source_sha: "",
          stagger,
        },
      };
    },
  };

  return grid;
}
